egs/wsj/s5/steps/libs/nnet3/train/dropout_schedule.py

```diff
@@ -0,0 +1,236 @@
+        raise
+
+    # reverse sort so that its easy to retrieve the dropout proportion
+    # for a particular data fraction
+    dropout_values.reverse()
+    for data_fraction, proportion in dropout_values:
+        assert data_fraction <= 1.0 and data_fraction >= 0.0
+        assert proportion <= 1.0 and proportion >= 0.0
+
+    return dropout_values
+
+
+def _get_component_dropout(dropout_schedule, data_fraction):
+    """Retrieve dropout proportion from schedule when data_fraction
+    proportion of data is seen. This value is obtained by using a
+    piecewise linear function on the dropout schedule.
+    This is a module-internal function called by _get_dropout_proportions().
+
+    See help for --trainer.dropout-schedule for how the dropout value
+    is obtained from the options.
+
+    Arguments:
+        dropout_schedule: A list of (data_fraction, dropout_proportion) values
+            sorted in descending order of data_fraction.
+        data_fraction: The fraction of data seen until this stage of
+            training.
+    """
+    if data_fraction == 0:
+        # Dropout at start of the iteration is in the last index of
+        # dropout_schedule
+        assert dropout_schedule[-1][0] == 0
+        return dropout_schedule[-1][1]
+    try:
+        # Find lower bound of the data_fraction. This is the
+        # lower end of the piecewise linear function.
+        (dropout_schedule_index, initial_data_fraction,
+         initial_dropout) = next((i, tup[0], tup[1])
+                                 for i, tup in enumerate(dropout_schedule)
+                                 if tup[0] <= data_fraction)
+    except StopIteration:
+        raise RuntimeError(
+            "Could not find data_fraction in dropout schedule "
+            "corresponding to data_fraction {0}.\n"
+            "Maybe something wrong with the parsed "
+            "dropout schedule {1}.".format(data_fraction, dropout_schedule))
+
+    if dropout_schedule_index == 0:
+        assert dropout_schedule[0][0] == 1 and data_fraction == 1
+        return dropout_schedule[0][1]
+
+    # The upper bound of data_fraction is at the index before the
+    # lower bound.
+    final_data_fraction, final_dropout = dropout_schedule[
+        dropout_schedule_index - 1]
+
+    if final_data_fraction == initial_data_fraction:
+        assert data_fraction == initial_data_fraction
+        return initial_dropout
+
+    assert (data_fraction >= initial_data_fraction
+            and data_fraction < final_data_fraction)
+
+    return ((data_fraction - initial_data_fraction)
+            * (final_dropout - initial_dropout)
+            / (final_data_fraction - initial_data_fraction)
+            + initial_dropout)
+
+
+def _get_dropout_proportions(dropout_schedule, data_fraction):
+    """Returns dropout proportions based on the dropout_schedule for the
+    fraction of data seen at this stage of training.
+    Returns None if dropout_schedule is None.
+
```

Comment:
Can you please give a couple of examples of what this function might return for different inputs, covering different types of input?  e.g. (and this will be wrong):
```
e.g.:
 _get_dropout_proportions('0.0,0.5,0.0', 0.75) = [ ('*', 0.75) ]
 _get_dropout_proportions('*=0.0,0.5,0.0,lstm.*=0.0,0.3@0.75,0.0', 0.75) = \
          [ ('*', 0.75), ('lstm.*', 0.3) ]
```
IMO it's always a good idea for this type of code to give such examples, it will
make maintenance much easier.  Please give examples for other functions in this
module; and remember to cover trivial cases such as where the input is the
empty string; there may be situations where 3 or 4 examples are needed to
demonstrate the function's range of behavior (but of course we'll
assume the reader is smart enough to extrapolate things).